Add JSpecify Nullability to the xml package

cppwfs opened the pull request:
* Tests were updated
* Test XML files updated to handle non nullable attributes and parameters
* In some cases I reformatted the whole xml file vs the changed section.   If this is a problem I'll undo.
* In many circumstances the chances of having null, should not happen, so `Objects.requireNonNull` was used.  But wonder if we should use Assert.state and have a message.   Thoughts?

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

cppwfs committed:
Add JSpecify Nullability to the xml package

* Tests were updated
* Test XML files updated to handle non nullable attributes and parameters
* In some cases I reformatted the whole xml file vs the changed section.   If this is a problem I'll undo.
* In many circumstances the chances of having null, should not happen, so `Objects.requireNonNull` was used.  But wonder if we should use Assert.state and have a message.   Thoughts?

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/DefaultInboundChannelAdapterParser.java:
```diff
@@ -46,8 +49,8 @@
-	@Override // NOSONAR complexity
-	protected BeanMetadataElement parseSource(Element element, ParserContext parserContext) { // NOSONAR
+	@Override
+	protected @Nullable BeanMetadataElement parseSource(Element element, ParserContext parserContext) {
```
Not-null. For same reason as before.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/DefaultInboundChannelAdapterParser.java:
```diff
@@ -76,6 +79,7 @@ protected BeanMetadataElement parseSource(Element element, ParserContext parserC
 								+ "element " + IntegrationNamespaceUtils.createElementDescription(element) + ".", source);
 				return null;
 			}
+			Assert.state(innerBeanDef != null, "No inner bean definition found"); //Required because JSpecify does not know how to use handle a preparsed boolean
```
I think `Objects.requireNonNull()` would read better.
We definitely never going to have it as null when we reach this place.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/DefaultInboundChannelAdapterParser.java:
```diff
@@ -91,6 +95,7 @@ else if (hasScriptElement) {
 								+ IntegrationNamespaceUtils.createElementDescription(element) + ".", source);
 				return null;
 			}
+			Assert.state(scriptElement != null, "No script element found");
```
Same `Objects.requireNonNull()` here instead.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/PointToPointChannelParser.java:
```diff
@@ -44,7 +45,7 @@
 
 	@Override
-	protected BeanDefinitionBuilder buildBeanDefinition(Element element, ParserContext parserContext) {
+	protected @Nullable BeanDefinitionBuilder buildBeanDefinition(Element element, ParserContext parserContext) {
```
As we discussed before: this is never null.

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/ChainParserTests-context.xml:
```diff
@@ -125,7 +125,7 @@
 	<chain id="subComponentsIdSupport1" input-channel="subComponentsIdSupport1Channel">
 		<splitter id="splitterWithinChain"/>
-		<resequencer id="resequencerWithinChain"/>
+		<resequencer id="resequencerWithinChain" correlation-strategy="errorChannel" release-strategy="errorChannel"/>
```
I think we need to revise the parser and avoid such a change in the XML configs.
See Java DSL for example:
```
	/**
	 * Populate the
	 * {@link org.springframework.integration.aggregator.ResequencingMessageHandler} with
	 * default options.
	 * @return the current {@link BaseIntegrationFlowDefinition}.
	 */
	public B resequence() {
```
so, having `<resequencer/>` in the XML is totally fine.
If that is otherwise, the respective XML parser has to be fixed.

artembilan reviewed (changes requested):
We can discuss my review off-line if you prefer.
Thanks

cppwfs committed:
Update based on code review

* Removed @Nullable from requested locations
* Added requested @SuppressWarning("NullAway") to specified locations
* Removed format changes from XML file that was mistakingly pushed.
* Resolved Nullability compile time errors from where the @Nullable was removed
* PointToPointChannelParser requires a null check on line 82 because builder because it may not have been set.   With this assert we can throw away the previous bogus exception

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/AbstractConsumerEndpointParser.java:
```diff
@@ -140,9 +142,9 @@ protected final AbstractBeanDefinition parseInternal(Element element, ParserCont
 		parserContext.registerBeanComponent(
-				new BeanComponentDefinition(handlerBeanDefinition, handlerBeanName, handlerAlias));
+				new BeanComponentDefinition(handlerBeanDefinition, handlerBeanName,
+						IntegrationNamespaceUtils.generateAlias(element)));
```
Correct, but the reason for such a local variable is to make the code more easier to read.

cppwfs committed:
Remove correlation and release strategy entries from  context-xml files
Remove CassandraParserUtilsTests.   Other tests provide the coverage required.
Readd method variable for readability

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/PointToPointChannelParser.java:
```diff
@@ -78,6 +78,7 @@ else if ((queueElement = DomUtils.getChildElementByTagName(element, "rendezvous-
 						"The 'fixed-subscriber' attribute is not allowed when a <queue/> child element is present.",
 						element);
 			}
+			Assert.state(builder != null, "builder must not be null");
```
Why not throw our friend "dummy" exception after error instead?
I mean this extra call is redundant and misleading.

cppwfs commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/AbstractChannelAdapterParser.java:
```diff
@@ -62,6 +63,7 @@ else if (!StringUtils.hasText(id)) {
 	}
 
 	@Override
+	@SuppressWarnings("NullAway")
```
This change is valid via our conversation.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/DefaultInboundChannelAdapterParser.java:
```diff
@@ -121,7 +119,7 @@ else if (hasRef) {
 			}
 		}
-		return result;
+		return Objects.requireNonNull(result);
```
Same here: why not exception after `error()`?

cppwfs replied:
There is a path where result is never set.   Unless we make the assumption that the check on line 118 can be removed.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/DefaultInboundChannelAdapterParser.java:
```diff
@@ -74,13 +76,12 @@ protected BeanMetadataElement parseSource(Element element, ParserContext parserC
 			}
 			else {
-				result = innerBeanDef.getBeanDefinition();
+				result = Objects.requireNonNull(innerBeanDef).getBeanDefinition();
```
DITTO

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/ChainParser.java:
```diff
@@ -142,14 +142,13 @@ private BeanMetadataElement parseChild(String chainHandlerId, Element element, i
 			BeanDefinition beanDefinition = parserContext.getDelegate().parseCustomElement(element, parentDefinition);
 			if (beanDefinition == null) {
 				parserContext.getReaderContext().error("child BeanDefinition must not be null", element);
-				return null;
```
Why not exception?

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/AbstractPollingInboundChannelAdapterParser.java:
```diff
@@ -60,7 +61,7 @@ else if (source instanceof RuntimeBeanReference runtimeBeanReference) {
 			parserContext.getReaderContext().error("Wrong 'source' type: must be 'BeanDefinition' or 'RuntimeBeanReference'", source);
 		}
-
+		Assert.state(sourceBeanName != null, "'sourceBeanName' must not be null");
```
Same here: exception after `error()` in the previous `else`

cppwfs committed:
Remove @Nullable from buildBeanDefinition in both AbstractChannelParser andk KafkaChannelParser

Add dummy exceptions to be applied below calls to .error so that NullAway knows that it will not return null
Make sure dummy exceptions have a consistent message describing why they are there

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/AbstractConsumerEndpointParser.java:
```diff
@@ -85,9 +87,9 @@ protected String getInputChannelAttributeName() {
+	protected final @Nullable AbstractBeanDefinition parseInternal(Element element, ParserContext parserContext) {
 		BeanDefinitionBuilder handlerBuilder = parseHandler(element, parserContext);
-		IntegrationNamespaceUtils.setValueIfAttributeDefined(handlerBuilder, element, "output-channel",
+		IntegrationNamespaceUtils.setValueIfAttributeDefined(Objects.requireNonNull(handlerBuilder), element, "output-channel",
```
This is not correct: the `parseHandler()` never returns `null`

artembilan commented on spring-integration-cassandra/src/main/java/org/springframework/integration/cassandra/config/xml/CassandraParserUtils.java:
```diff
@@ -91,7 +92,7 @@ public static void processOutboundTypeAttributes(Element element, ParserContext
 	}
 
-	public static boolean areMutuallyExclusive(String query, BeanDefinition statementExpressionDef,
+	public static boolean areMutuallyExclusive(String query, @Nullable BeanDefinition statementExpressionDef,
```
And why this method is not `private` yet since you have already removed that redundant test?

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/AbstractPollingInboundChannelAdapterParser.java:
```diff
@@ -59,8 +59,8 @@ else if (source instanceof RuntimeBeanReference runtimeBeanReference) {
 		}
 		else {
 			parserContext.getReaderContext().error("Wrong 'source' type: must be 'BeanDefinition' or 'RuntimeBeanReference'", source);
+			throw new IllegalStateException("This dummy exception is meant to signify to  NullAway that an the error method throws an exception");
```
I don't think it is correct and polite to say that as an exception message.
Just copy/paste from `error()` and be with it!
Plus, keep in mind the rule of not more than 120 symbols in line.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/AbstractPollingInboundChannelAdapterParser.java:
```diff
@@ -59,8 +59,8 @@ else if (source instanceof RuntimeBeanReference runtimeBeanReference) {
 			parserContext.getReaderContext().error("Wrong 'source' type: must be 'BeanDefinition' or 'RuntimeBeanReference'", source);
+			throw new IllegalStateException("This dummy exception is meant to signify to  NullAway that an the error method throws an exception");
 		}
-
```
Please, bring this line back.
The code should be as readable as possible.

cppwfs committed:
Add blank line to improve readability in `AbstractPollingInboundChannelAdapterParser` class.

Replace stock exception for NullAway to signal that `error` will throw an exception.
  With the message passed to the `error` method.
